tool/tsh/gcp.go

```diff
@@ -126,9 +128,47 @@ func (a *gcpApp) Close() error {
 	return trace.NewAggregate(errs...)
 }
 
+func (a *gcpApp) getGcloudConfigPath() string {
+	return path.Join(profile.FullProfilePath(a.cf.HomePath), "gcp", a.app.ClusterName, a.app.Name, "gcloud")
+}
+
+func projectIDFromServiceAccountName(serviceAccount string) (string, error) {
+	if serviceAccount == "" {
+		return "", trace.BadParameter("invalid service account format: empty string received")
+	}
+
+	user, domain, found := strings.Cut(serviceAccount, "@")
+	if !found {
+		return "", trace.BadParameter("invalid service account format: missing @")
+	}
+	if user == "" {
+		return "", trace.BadParameter("invalid service account format: empty user")
+	}
+
+	projectID, iamDomain, found := strings.Cut(domain, ".")
+	if !found {
+		return "", trace.BadParameter("invalid service account format: missing <project-id>.iam.gserviceaccount.com after @")
+	}
+
+	if projectID == "" {
+		return "", trace.BadParameter("invalid service account format: missing project ID")
+	}
+
+	if iamDomain != "iam.gserviceaccount.com" {
+		return "", trace.BadParameter("invalid service account format: expected suffix %q, got %q", "iam.gserviceaccount.com", iamDomain)
```

Comment:
nit: move "iam.gserviceaccount.com" into a constant somewhere? Only because we reference that exact string 3 times

Reply:
I'm doing some refactoring for the upcoming PR dealing with additional validation of GCP service accounts in `tctl` and `tsh app login`, I'll move it to constant there.

Reply:
Done in https://github.com/gravitational/teleport/pull/20180.